OptimizelySDK/Event/BatchEventProcessor.cs

```diff
@@ -0,0 +1,358 @@
+            {
+                Logger.Log(LogLevel.INFO, "Exiting processing loop. Attempting to flush pending events.");
+                FlushQueue();
+            }
+        }
+
+        public void Flush()
+        {
+            EventQueue.Add(FLUSH_SIGNAL);
+        }
+
+        private void FlushQueue()
+        {
+            if (CurrentBatch.Count == 0)
+            {
+                return;
+            }
+
+            List<UserEvent> toProcessBatch = null;
+            lock (mutex)
+            {
+                toProcessBatch = new List<UserEvent>(CurrentBatch);
+                CurrentBatch.Clear();
+            }
+            
+
+            LogEvent logEvent = EventFactory.CreateLogEvent(toProcessBatch.ToArray(), Logger);
+
+            NotificationCenter?.SendNotifications(NotificationCenter.NotificationType.LogEvent, logEvent);
+
+            try
+            {
+                EventDispatcher?.DispatchEvent(logEvent);
+            }
+            catch (Exception e)
+            {
+                Logger.Log(LogLevel.ERROR, "Error dispatching event: " + logEvent + " " + e);
+            }
+        }
+
+        /// <summary>
+        /// Stops batch event processor.
+        /// </summary>
+        public void Stop()
+        {
+            if (Disposed) return;
+
+            EventQueue.Add(SHUTDOWN_SIGNAL);
+            
+            if (!Executer.Join(TimeoutInterval))
+                Logger.Log(LogLevel.ERROR, $"Timeout exceeded attempting to close for {TimeoutInterval.Milliseconds} ms");
+
+            IsStarted = false;
+            Logger.Log(LogLevel.WARN, $"Stopping scheduler.");
+        }
+        
+        public void Process(UserEvent userEvent) {
+            Logger.Log(LogLevel.DEBUG, "Received userEvent: " + userEvent);
+
+            if (Disposed) { 
+                Logger.Log(LogLevel.WARN, "Executor shutdown, not accepting tasks.");
+                return;
+            }
+
+            if (EventQueue.TryAdd(userEvent))
+            {
+                Logger.Log(LogLevel.WARN, "Payload not accepted by the queue.");
+            }
+        }
+        
+        private void AddToBatch(UserEvent userEvent)
+        {
+            if (ShouldSplit(userEvent))
+            {
+                FlushQueue();
+                CurrentBatch = new List<UserEvent>();
+            }
+
+            // Reset the deadline if starting a new batch.
+            if (CurrentBatch.Count == 0)
+                FlushingIntervalDeadline = DateTime.Now.Millisecond + FlushInterval.Milliseconds;
+
+            lock (mutex) {
+                CurrentBatch.Add(userEvent);
+            }
+            
+            if (CurrentBatch.Count >= BatchSize) {
+                FlushQueue();
+            }
+        }
+
+        private bool ShouldSplit(UserEvent userEvent) {
+            if (CurrentBatch.Count == 0) {
+                return false;
+            }
+
+            EventContext currentContext = CurrentBatch.Last().Context;
```

Comment:
Do we need to lock here since we are reading from the CurrentBatch?

Reply:
Added lock statement.